File: apollo-integration/src/test/java/com/apollographql/apollo/SendOperationIdentifiersTest.java
package com.apollographql.apollo;

import com.apollographql.apollo.api.Input;
import com.apollographql.apollo.integration.normalizer.HeroAndFriendsNamesQuery;

import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.concurrent.atomic.AtomicBoolean;

import io.reactivex.functions.Predicate;
import okhttp3.Dispatcher;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.mockwebserver.MockWebServer;

import static com.apollographql.apollo.Utils.enqueueAndAssertResponse;
import static com.apollographql.apollo.integration.normalizer.type.Episode.EMPIRE;
import static com.google.common.truth.Truth.assertThat;

public class SendOperationIdentifiersTest {
  @Rule public final MockWebServer server = new MockWebServer();

  @Test public void sendOperationIdsTrue() throws Exception {
    final HeroAndFriendsNamesQuery query = new HeroAndFriendsNamesQuery(Input.fromNullable(EMPIRE));
    ApolloClient apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .enableAutoPersistedQueries(true)
        .build();
    apolloClient.query(query).enqueue(null);

    String serverRequest = server.takeRequest().getBody().readUtf8();
    assertThat(serverRequest.contains(String.format("\"sha256Hash\":\"%s\"", query.operationId()))).isTrue();
    assertThat(serverRequest.contains("\"query\":")).isFalse();
  }

  @Test public void doesNotSendOperationIdsWhenFalse() throws Exception {
    final HeroAndFriendsNamesQuery query = new HeroAndFriendsNamesQuery(Input.fromNullable(EMPIRE));
    ApolloClient apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .enableAutoPersistedQueries(false)
        .build();
    apolloClient.query(query).enqueue(null);

    String serverRequest = server.takeRequest().getBody().readUtf8();
    assertThat(serverRequest.contains("\"id\":\"")).isFalse();
    assertThat(serverRequest.contains("\"query\":")).isTrue();
  }

  @Test public void operationIdHttpRequestHeader() throws Exception {
    final HeroAndFriendsNamesQuery heroAndFriendsNamesQuery = new HeroAndFriendsNamesQuery(Input.fromNullable(EMPIRE));
    final AtomicBoolean applicationInterceptorHeader = new AtomicBoolean();
    final AtomicBoolean networkInterceptorHeader = new AtomicBoolean();
    OkHttpClient okHttpClient = new OkHttpClient.Builder()
        .dispatcher(new Dispatcher(Utils.immediateExecutorService()))
        .addInterceptor(new Interceptor() {
          @Override public okhttp3.Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            if (request.header("X-APOLLO-OPERATION-ID").equals(heroAndFriendsNamesQuery.operationId())) {
              applicationInterceptorHeader.set(true);
            }
            return chain.proceed(chain.request());
          }
        })
        .addNetworkInterceptor(new Interceptor() {
          @Override public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            if (request.header("X-APOLLO-OPERATION-ID").equals(heroAndFriendsNamesQuery.operationId())) {
              networkInterceptorHeader.set(true);
            }
            return chain.proceed(chain.request());
          }
        })
        .build();

    ApolloClient apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient)
        .dispatcher(Utils.immediateExecutor())
        .build();

    enqueueAndAssertResponse(
        server,
        "HeroAndFriendsNameResponse.json",
        apolloClient.query(heroAndFriendsNamesQuery),
        new Predicate<com.apollographql.apollo.api.Response<HeroAndFriendsNamesQuery.Data>>() {
          @Override
          public boolean test(com.apollographql.apollo.api.Response<HeroAndFriendsNamesQuery.Data> response) throws Exception {
            return !response.hasErrors();
          }
        }
    );

    assertThat(applicationInterceptorHeader.get()).isTrue();
    assertThat(networkInterceptorHeader.get()).isTrue();
  }
}

File: apollo-integration/src/test/kotlin/com/apollographql/apollo3/SendOperationIdentifiersTest.kt
package com.apollographql.apollo3

import com.apollographql.apollo3.Utils.enqueueAndAssertResponse
import com.apollographql.apollo3.Utils.immediateExecutor
import com.apollographql.apollo3.Utils.immediateExecutorService
import com.apollographql.apollo3.api.Input
import com.apollographql.apollo3.api.Response
import com.apollographql.apollo3.integration.normalizer.HeroAndFriendsNamesQuery
import com.apollographql.apollo3.integration.normalizer.type.Episode
import com.google.common.truth.Truth
import io.reactivex.functions.Predicate
import okhttp3.Dispatcher
import okhttp3.OkHttpClient
import okhttp3.mockwebserver.MockWebServer
import org.junit.Rule
import org.junit.Test
import java.util.concurrent.atomic.AtomicBoolean

class SendOperationIdentifiersTest {
  val server = MockWebServer()
  @Test
  @Throws(Exception::class)
  fun sendOperationIdsTrue() {
    val query = HeroAndFriendsNamesQuery(Input.present(Episode.EMPIRE))
    val apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .enableAutoPersistedQueries(true)
        .build()
    apolloClient.query(query).enqueue(null)
    val serverRequest = server.takeRequest().body.readUtf8()
    Truth.assertThat(serverRequest.contains("extensions")).isTrue()
    Truth.assertThat(serverRequest.contains("persistedQuery")).isTrue()
    Truth.assertThat(serverRequest.contains(String.format("\"sha256Hash\":\"%s\"", query.operationId()))).isTrue()
    Truth.assertThat(serverRequest.contains("\"query\":")).isFalse()
  }

  @Test
  @Throws(Exception::class)
  fun doesNotSendOperationIdsWhenFalse() {
    val query = HeroAndFriendsNamesQuery(Input.present(Episode.EMPIRE))
    val apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .enableAutoPersistedQueries(false)
        .build()
    apolloClient.query(query).enqueue(null)
    val serverRequest = server.takeRequest().body.readUtf8()
    Truth.assertThat(serverRequest.contains("extensions")).isFalse()
    Truth.assertThat(serverRequest.contains("persistedQuery")).isFalse()
    Truth.assertThat(serverRequest.contains("sha256Hash")).isFalse()
    Truth.assertThat(serverRequest.contains("\"query\":")).isTrue()
  }

  @Test
  @Throws(Exception::class)
  fun operationIdHttpRequestHeader() {
    val heroAndFriendsNamesQuery = HeroAndFriendsNamesQuery(Input.present(Episode.EMPIRE))
    val applicationInterceptorHeader = AtomicBoolean()
    val networkInterceptorHeader = AtomicBoolean()
    val okHttpClient = OkHttpClient.Builder()
        .dispatcher(Dispatcher(immediateExecutorService()))
        .addInterceptor { chain ->
          val request = chain.request()
          if (request.header("X-APOLLO-OPERATION-ID") == heroAndFriendsNamesQuery.operationId()) {
            applicationInterceptorHeader.set(true)
          }
          chain.proceed(chain.request())
        }
        .addNetworkInterceptor { chain ->
          val request = chain.request()
          if (request.header("X-APOLLO-OPERATION-ID") == heroAndFriendsNamesQuery.operationId()) {
            networkInterceptorHeader.set(true)
          }
          chain.proceed(chain.request())
        }
        .build()
    val apolloClient = ApolloClient.builder()
        .serverUrl(server.url("/"))
        .okHttpClient(okHttpClient)
        .dispatcher(immediateExecutor())
        .build()
    enqueueAndAssertResponse(
        server,
        "HeroAndFriendsNameResponse.json",
        apolloClient.query(heroAndFriendsNamesQuery)
    ) { response -> !response.hasErrors() }
    Truth.assertThat(applicationInterceptorHeader.get()).isTrue()
    Truth.assertThat(networkInterceptorHeader.get()).isTrue()
  }
}
